### PR TITLE
fix nestedPendingOperations mount and umount parallel bug

### LINKED ISSUE #109047: kubelet nestedPendingOperations may leak operation lead same pv not to do mount or umount operation
### What happened?

on my scenes，create same cronjob use same sfs(the nfs storage on Huawei Cloud) type storage by pvc to do same training work. At the same time, a node will schedule many pods using the same pvc, and there will be many mount and umount tasks about this pvc executed at the same time. And we find all pods on a node at container creating state, because of waiting pvc to mount.
check log, we can't find any we can't find any operationExecutor.MountVolume started or failed log like "operationExecutor.MountVolume started for"

and also we check  volume_manager_total_volumes metrics, volume is add to dsw.
analyze code, if isExpectedError, kubelet not log any thing
``` golang 
// ignore nestedpendingoperations.IsAlreadyExists and exponentialbackoff.IsExponentialBackoff errors, they are expected.
func isExpectedError(err error) bool {
	return nestedpendingoperations.IsAlreadyExists(err) || exponentialbackoff.IsExponentialBackoff(err)
}
```
and operationExecutor will generate operation, add call pendingOperations Run, pendingOperations Run will check operation slice, to find whether has same operation exists, and if operation exist, will check operation's operationPending status, if equal true, means same operation now is excuting, and return AlreadyExists.
operationKey composed of volumeName, podName, nodeName
``` golang 
// MountVolume Func
// Avoid executing mount/map from multiple pods referencing the
// same volume in parallel
podName := nestedpendingoperations.EmptyUniquePodName

// TODO: remove this -- not necessary
if !volumeToMount.PluginIsAttachable && !volumeToMount.PluginIsDeviceMountable {
	// volume plugins which are Non-attachable and Non-deviceMountable can execute mount for multiple pods
	// referencing the same volume in parallel
	podName = util.GetUniquePodName(volumeToMount.Pod)
}

// TODO mount_device
return oe.pendingOperations.Run(
		volumeToMount.VolumeName, podName, "" /* nodeName */, generatedOperations)
```
so MountVolume operationKey {volumeName, EmptyUniquePodName,EmptyNodeName}, use EmptyUniquePodName to avoid executing mount/map from multiple pods referencing the same volume in parallel

``` golang 
// UnmountVolume Func
// All volume plugins can execute unmount/unmap for multiple pods referencing the
	// same volume in parallel
	podName := volumetypes.UniquePodName(volumeToUnmount.PodUID)

	return oe.pendingOperations.Run(
		volumeToUnmount.VolumeName, podName, "" /* nodeName */, generatedOperations)
```
soUnmountVolume operationKey {volumeName, podUID,EmptyNodeName}, because All volume plugins can execute unmount/unmap for multiple pods referencing the same volume in parallel

now see isOperationExists func, if previousOp.podName equals to EmptyUniquePodName  or current operation podName  equals to EmptyUniquePodName, will return podNameMatch, mount operation can find  umount operation or umount operation can find mount operation
``` golang 
// This is an internal function and caller should acquire and release the lock
func (grm *nestedPendingOperations) isOperationExists(key operationKey) (bool, int) {

	// If volumeName is empty, operation can be executed concurrently
	if key.volumeName == EmptyUniqueVolumeName {
		return false, -1
	}

	for previousOpIndex, previousOp := range grm.operations {
		volumeNameMatch := previousOp.key.volumeName == key.volumeName

		podNameMatch := previousOp.key.podName == EmptyUniquePodName ||
			key.podName == EmptyUniquePodName ||
			previousOp.key.podName == key.podName

		nodeNameMatch := previousOp.key.nodeName == EmptyNodeName ||
			key.nodeName == EmptyNodeName ||
			previousOp.key.nodeName == key.nodeName

		if volumeNameMatch && podNameMatch && nodeNameMatch {
			return true, previousOpIndex
		}
	}

	return false, -1
}
``` 

``` golang 
func (grm *nestedPendingOperations) Run(
	volumeName v1.UniqueVolumeName,
	podName volumetypes.UniquePodName,
	nodeName types.NodeName,
	generatedOperations volumetypes.GeneratedOperations) error {
	grm.lock.Lock()
	defer grm.lock.Unlock()

	opKey := operationKey{volumeName, podName, nodeName}

	opExists, previousOpIndex := grm.isOperationExists(opKey)
	if opExists {
		previousOp := grm.operations[previousOpIndex]
		// Operation already exists
		if previousOp.operationPending {
			// Operation is pending
			return NewAlreadyExistsError(opKey)
		}

		backOffErr := previousOp.expBackoff.SafeToRetry(fmt.Sprintf("%+v", opKey))
		if backOffErr != nil {
			if previousOp.operationName == generatedOperations.OperationName {
				return backOffErr
			}
			// previous operation and new operation are different. reset op. name and exp. backoff
			grm.operations[previousOpIndex].operationName = generatedOperations.OperationName
			grm.operations[previousOpIndex].expBackoff = exponentialbackoff.ExponentialBackoff{}
		}

		// Update existing operation to mark as pending.
		grm.operations[previousOpIndex].operationPending = true
		grm.operations[previousOpIndex].key = opKey
	} else {
		// Create a new operation
		grm.operations = append(grm.operations,
			operation{
				key:              opKey,
				operationPending: true,
				operationName:    generatedOperations.OperationName,
				expBackoff:       exponentialbackoff.ExponentialBackoff{},
			})
	}

	go func() (eventErr, detailedErr error) {
		// Handle unhandled panics (very unlikely)
		defer k8sRuntime.HandleCrash()
		// Handle completion of and error, if any, from operationFunc()
		defer grm.operationComplete(opKey, &detailedErr)
		return generatedOperations.Run()
	}()

	return nil
}
```  
because of isOperationExists, this mean if a volume now is excuting mount, a umount can't add to operations, and if a umount  now is excuting, a mount operation can't add to operations, but  multiple pods referencing the same volume can add multiple umount operation into operation, may lead operation leak.
```
t0 Operations: {umount, pvname, pod1, pending=true} {umount, pvname, pod2, pending=true}
t1 Operations: {umount, pvname, pod1, pending=false} {umount, pvname, pod2, pending=true}
t2 Operations: {mount, pvname, EmptyUniquePodName, pending=true} {umount, pvname, pod2, pending=true}
t3 Operations: {mount, pvname, EmptyUniquePodName, pending=false} {umount, pvname, pod2, pending=true}
t4 Operations: {umount, pvname, pod2, pending=true} {umount, pvname, pod2, pending=true}
t5 Operations: {umount, pvname, pod2, pending=false} {umount, pvname, pod2, pending=true}
t6 Operations: {umount, pvname, pod2, pending=true}
```
t0 two pod do umount volume operation
t1 pod1 umount failed,  update pending to false
t2 pod3 add, operation excuter add mount operation, will use index 0, override umount operation
t3 mount operation failed , update pending to false
t4 pod2 umount operation add, will find index 0, add override opertation, now ,we have two same umount operation, now we have two goroutines to do pod2 umount operations
t5 first goroutine pod2 umount failed, will update index 0 operation, update pending to false
t6 second goroutine pod2 umount success, will delete index 0 operation,
finally lead a umount opertion in cache, and this operation will lead all mount opeartion  return AlreadyExistsError, and not cause any csi call.

and i think there  exists some other scene can trigger this leak.

and also, when operation success, deleteOperation, will use tail element to replace, and may increase the probalility of occurrence. because isOperationExists will always pick first match one.
```
func (grm *nestedPendingOperations) deleteOperation(key operationKey) {
	// Assumes lock has been acquired by caller.

	opIndex := -1
	for i, op := range grm.operations {
		if op.key.volumeName == key.volumeName &&
			op.key.podName == key.podName &&
			op.key.nodeName == key.nodeName {
			opIndex = i
			break
		}
	}

	if opIndex < 0 {
		return
	}

	// Delete index without preserving order
	grm.operations[opIndex] = grm.operations[len(grm.operations)-1]
	grm.operations = grm.operations[:len(grm.operations)-1]
}
```

and also override  mechanism may lead exponentialBackOff invalid， for example mount and umount take turns to excute.

### What did you expect to happen?

pv can mount , pod can running

### How can we reproduce it (as minimally and precisely as possible)?

reproduce step:
create same cronjob that use same pvc(use csi sfs)
and csi driver restart periodic

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>
1.23

### Cloud provider

<details>
HuaweiCloud
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>
CSI

### PR DESCRIPTION
**Pull request by 249043822**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/109047

#### Special notes for your reviewer:
/cc @jingxu97 @gnufied 
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by yangjunmyfm192085 — COMMENTED** [review]
/sig node

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go [hunk 1]
```diff
@@ -256,10 +257,18 @@ func (grm *nestedPendingOperations) isOperationExists(key operationKey) (bool, i
 			previousOp.key.nodeName == key.nodeName
 
 		if volumeNameMatch && podNameMatch && nodeNameMatch {
-			return true, previousOpIndex
+			if previousOp.operationPending {
+				return true, previousOpIndex
+			} else {
+				if opIndex == -1 {
+					opIndex = previousOpIndex
+				}
```
[the remark this answers is not in the document]
Comment (249043822) [reply pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:265]: Thanks for reivew, missing exact match first, I will add a test case for this.

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go [hunk 2]
```diff
@@ -256,10 +257,21 @@ func (grm *nestedPendingOperations) isOperationExists(key operationKey) (bool, i
...
+			if previousOp.operationPending {
+				return true, previousOpIndex
+			}
+			// then exact match prior non-exact match
+			podNameExactMatch := previousOp.key.podName == key.podName
+			nodeNameExactMatch := previousOp.key.nodeName == key.nodeName
+			if opIndex == -1 || (podNameExactMatch && nodeNameExactMatch) {
+				opIndex = previousOpIndex
+			}
 		}
```
Comment (gnufied) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:270]: Thanks for making the fix. Does following logic seems more direct?

```golang
        exactMatch := -1
	nonExactMatch := -1

	for previousOpIndex, previousOp := range grm.operations {
		volumeNameMatch := previousOp.key.volumeName == key.volumeName

		if volumeNameMatch {
			if key.podName == previousOp.key.podName && key.nodeName == previousOp.key.nodeName {
				exactMatch = previousOpIndex
				break
			}

			podNameMatch := previousOp.key.podName == EmptyUniquePodName ||
				key.podName == EmptyUniquePodName ||
				previousOp.key.podName == key.podName

			nodeNameMatch := previousOp.key.nodeName == EmptyNodeName ||
				key.nodeName == EmptyNodeName ||
				previousOp.key.nodeName == key.nodeName

			if podNameMatch && nodeNameMatch {
				nonExactMatch = previousOpIndex
			}
		}
	}
	if exactMatch != -1 {
		return true, exactMatch
	}
	if nonExactMatch != -1 {
		return true, nonExactMatch
	}
```

Basically we want to pick exactly matching operation over other ones.

File: test/e2e/storage/persistent_volumes.go [hunk 3]
```diff
@@ -41,15 +45,15 @@ import (
 
 // Validate PV/PVC, create and verify writer pod, delete the PVC, and validate the PV's
 // phase. Note: the PV is deleted in the AfterEach, not here.
-func completeTest(f *framework.Framework, c clientset.Interface, ns string, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim) {
+func completeTest(f *framework.Framework, c clientset.Interface, ns string, pv *v1.PersistentVolume, pvc *v1.PersistentVolumeClaim, act func(c clientset.Interface, t *framework.TimeoutContext, ns string, pvc *v1.PersistentVolumeClaim, command string) error) {
 	// 1. verify that the PV and PVC have bound correctly
```
Comment (gnufied) [inline test/e2e/storage/persistent_volumes.go:49]: Does this test case reproduces the issue reliably?

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go (hunk 2 above)
Comment (gnufied) [reply pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:270]: Well the code does not return first pending match though. One downside of current approach now is - we must iterate through all operations before `IsOperationExists` can return. This could potentially be expensive on a busy node.

File: test/e2e/storage/persistent_volumes.go (hunk 3 above)
Comment (Dingshujie) [reply test/e2e/storage/persistent_volumes.go:49]: hard to say.  some operation orders required to reproduce the problem, this test case can't guarantee it.

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go (hunk 2 above)
Comment (249043822) [reply pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:270]: Yea, in worst case, there is no pending match operations at all but there is a non exact match at the header, we must still iterate all. But I think we don't change the time complexity of O(N) from global perspective, because we should always iterate the slice to find a match volume operation if there are many different volumes operations. shall we add a TODO to improve the algorithm using other data structure in future?

File: test/e2e/storage/persistent_volumes.go (hunk 3 above)
Comment (249043822) [reply test/e2e/storage/persistent_volumes.go:49]: Agree, not reliable, only add a e2e test coverage to prevent regression

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go (hunk 2 above)
Comment (gnufied) [reply pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:270]: Yeah I wonder if we can change the data structure used here from a slice to something that can accommodate these requirements in a better way. The code is also kinda hard to follow because of underlying complexity (not this PR's fault, it was always hard to follow).

Comment (Dingshujie) [reply pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:270]: aggree,  need to refactor this data structure. 

proposing a kep to do this?

**Comment by 249043822** [comment]
@gnufied  All the tests are passed, but i have a little nit, while volumeName is empty, which is for verify_volumes_are_attached_per_node operation,  if the operation for the same node is still pending,  shall we reject current operation for the same node?  
From the test case https://github.com/kubernetes/kubernetes/blob/0424c7c74d926b4fe3193059e003e9056b429d28/pkg/volume/util/operationexecutor/operation_executor_test.go#L191, we can know that verify_volumes_are_attached_per_node for the same node can run conccurently

**Comment by jingxu97** [comment]
@249043822 [249043822](https://github.com/249043822), thanks a lot for finding the issue and working on the fix.
I was also thinking about this logic and propose a fix here https://github.com/kubernetes/kubernetes/pull/109343, the change is smaller. Will that fix logic work for your case?

**Comment by 249043822** [comment]
> @249043822 [249043822](https://github.com/249043822), thanks a lot for finding the issue and working on the fix. I was also thinking about this logic and propose a fix here #109343, the change is smaller. Will that fix logic work for your case?

@jingxu97 thanks for your reply, the fix in #109343 is similar with the first version which i have committed, it's really a minimal change for the bug. Talking with @gnufied ,  Hemant thinks that  fix has one downside - we must iterate through all operations before IsOperationExists can return. This could potentially be expensive on a busy node. https://github.com/kubernetes/kubernetes/pull/109190#discussion_r840819369

so i refactor a new fix as this

File: pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go [hunk 4]
```diff
@@ -319,36 +278,185 @@ func (grm *nestedPendingOperations) operationComplete(key operationKey, err *err
...
+	// Assumes lock has been acquired by caller.
+
+	vro, exists := o.pendingOps[volumeName]
+	if !exists {
+		return false
+	}
+	if volumeName == EmptyUniqueVolumeName {
+		// volumeName empty, nodeName should exists(for verify_volumes_are_attached_per_node)
+		// to make the opKey unique. if the operation for the same node is still pending,
+		// should reject current operation?
```
Comment (jingxu97) [inline pkg/volume/util/nestedpendingoperations/nestedpendingoperations.go:426]: right now only verifyVolumeAttached function uses empty volumeName to allow it to run parallel no matter what other volume operations are running. I think we can still make this if part at the beginning of the function

**Comment by Dingshujie** [comment]
@SergeyKanzhelev @jingxu97 can you reviewed this pr？

**Comment by xhejtman** [comment]
Hello, is there any progress here? The bug is actually a big problem for nextflow computation as tasks get stucked in container creating..

**Comment by 249043822** [comment]
@gnufied @jingxu97 @Dingshujie As this refactor is a big change and may not make sense to all, so can we make this work two steps? First we can consider the smaller change in: https://github.com/kubernetes/kubernetes/pull/110951 to be landed to solve this bug, then we may promote a nice refactor change for this in next step to make the work completely.
